public/app/features/dashboard-scene/scene/DashboardScene.test.tsx

```diff
@@ -905,8 +906,229 @@ describe('DashboardScene', () => {
       expect(scene.managedResourceCannotBeEdited()).toBe(false);
     });
   });
+
+  describe('getExpressionTypes', () => {
+    it('should extract SQL expression type from V1 and V2 dashboards', () => {
+      const scene = buildTestScene();
+      const saveModel = createV1DashboardWithExpressions(['sql']);
+      const saveModelV2 = createV2DashboardWithExpressions(['sql']);
+
+      const result = scene.getExpressionTypes(saveModel);
+      const resultV2 = scene.getExpressionTypes(saveModelV2);
+
+      expect(result).toEqual(['sql']);
+      expect(resultV2).toEqual(['sql']);
+    });
+
+    it('should extract multiple expression types from V1 and V2 dashboards', () => {
+      const scene = buildTestScene();
+      const saveModel = createV1DashboardWithExpressions(['sql', 'reduce', 'math']);
+      const saveModelV2 = createV2DashboardWithExpressions(['sql', 'reduce', 'math']);
+
+      const result = scene.getExpressionTypes(saveModel);
+      const resultV2 = scene.getExpressionTypes(saveModelV2);
+
+      expect(result).toEqual(['sql', 'reduce', 'math']);
+      expect(resultV2).toEqual(['sql', 'reduce', 'math']);
+    });
+
+    it('should deduplicate expression types', () => {
+      const scene = buildTestScene();
+      const saveModel = createV1DashboardWithExpressions(['sql', 'sql', 'reduce']);
+      const saveModelV2 = createV2DashboardWithExpressions(['sql', 'sql', 'reduce']);
+
+      const result = scene.getExpressionTypes(saveModel);
+      const resultV2 = scene.getExpressionTypes(saveModelV2);
+
+      expect(result).toEqual(['sql', 'reduce']);
+      expect(resultV2).toEqual(['sql', 'reduce']);
+    });
+
+    it('should return undefined when no expressions exist for V1 and V2 dashboards', () => {
+      const scene = buildTestScene();
+      const saveModel = createV1DashboardWithExpressions([]);
+      const saveModelV2 = createV2DashboardWithExpressions([]);
+
+      const result = scene.getExpressionTypes(saveModel);
+      const resultV2 = scene.getExpressionTypes(saveModelV2);
+
+      expect(result).toBeUndefined();
+      expect(resultV2).toBeUndefined();
+    });
+
+    it('should return undefined for empty dashboard', () => {
+      const scene = buildTestScene();
+      const saveModel = { panels: [] } as unknown as Dashboard;
+
+      const result = scene.getExpressionTypes(saveModel);
+
+      expect(result).toBeUndefined();
+    });
+
+    it('should skip non-expression datasources', () => {
+      const scene = buildTestScene();
+      const saveModel = {
+        panels: [
+          {
+            type: 'timeseries',
+            targets: [
+              {
+                datasource: { type: 'prometheus', uid: 'prometheus-uid' },
+                type: 'instant',
+              },
+            ],
+          },
+        ],
+      } as unknown as Dashboard;
+
+      const result = scene.getExpressionTypes(saveModel);
+
+      expect(result).toBeUndefined();
```

Comment:
@alexjonspencer1 maybe this is idiomatic, but somehow I had expected to see an empty array here. Any thoughts on this?